File: scripts/classification_rules.py
from __future__ import annotations

import json
import re
from pathlib import Path


RULES_PATH = Path(__file__).with_name("classification_rules.json")
CLASSIFICATION_RULES = json.loads(RULES_PATH.read_text(encoding="utf-8"))


def _normalize_text(value: str | None) -> str:
    return (value or "").lower()


def _build_combined_text(post: dict) -> str:
    return f"{_normalize_text(post.get('title'))} {_normalize_text(post.get('searchText'))}".strip()


def _matches_any(text: str, needles: list[str]) -> bool:
    return any(needle in text for needle in needles)


def infer_setup_candidates(text: str) -> list[str]:
    lower = _normalize_text(text)
    candidates = [
        setup
        for setup, needles in CLASSIFICATION_RULES["setupCandidates"].items()
        if _matches_any(lower, needles)
    ]
    return candidates or ["uncategorized"]


def generate_normalized_tags(post: dict) -> dict:
    combined = _build_combined_text(post)
    labels = post.get("labels") or []
    candidates = post.get("setupCandidates") or []
    defaults = CLASSIFICATION_RULES["defaults"]

    post_id = post.get("id", "")
    if post_id.startswith("nt-page-") or "static-page" in labels:
        content_type = "reference"
    elif re.match(r"^nt-\d{4}-\d{2}-\d{2}-", post_id):
        content_type = (
            "theory"
            if _matches_any(_normalize_text(post.get("title")), CLASSIFICATION_RULES["contentType"]["seriesKeywords"])
            else "daily"
        )
    else:
        content_type = defaults["content_type"]

    market = [
        tag
        for tag, needles in CLASSIFICATION_RULES["normalizedTags"]["market"].items()
        if _matches_any(combined, needles)
    ]

    setup = []
    for tag, config in CLASSIFICATION_RULES["normalizedTags"]["setup"].items():
        candidate_match = any(candidate in candidates for candidate in config.get("candidates", []))
        if candidate_match or _matches_any(combined, config.get("needles", [])):
            setup.append(tag)

    topic = [
        tag
        for tag, needles in CLASSIFICATION_RULES["normalizedTags"]["topic"].items()
        if _matches_any(combined, needles)
    ]

    if _matches_any(combined, CLASSIFICATION_RULES["normalizedTags"]["level"]["beginner"]):
        level = "beginner"
    elif _matches_any(combined, CLASSIFICATION_RULES["normalizedTags"]["level"]["advanced"]):
        level = "advanced"
    else:
        level = defaults["level"]

    return {
        "content_type": content_type,
        "market": market or [defaults["market"]],
        "setup": setup or [defaults["setup"]],
        "topic": topic or [defaults["topic"]],
        "level": level,
    }

File: scripts/classification-rules.mjs
import { readFileSync } from "fs";

const rulesPath = new URL("./classification_rules.json", import.meta.url);

export const classificationRules = JSON.parse(readFileSync(rulesPath, "utf-8"));

function normalizeText(value) {
  return String(value || "").toLowerCase();
}

function buildCombinedText(post) {
  return `${normalizeText(post.title)} ${normalizeText(post.searchText)}`.trim();
}

function matchesAny(text, needles = []) {
  return needles.some((needle) => text.includes(needle));
}

export function inferSetupCandidates(input) {
  const text = typeof input === "string" ? normalizeText(input) : buildCombinedText(input);
  const candidates = [];

  for (const [setup, needles] of Object.entries(classificationRules.setupCandidates)) {
    if (matchesAny(text, needles)) {
      candidates.push(setup);
    }
  }

  return candidates.length > 0 ? candidates : ["uncategorized"];
}

export function generateNormalizedTags(post) {
  const combined = buildCombinedText(post);
  const labels = post.labels || [];
  const candidates = post.setupCandidates || [];

  let contentType = classificationRules.defaults.content_type;
  if (post.id?.startsWith("nt-page-") || labels.includes("static-page")) {
    contentType = "reference";
  } else if (/^nt-\d{4}-\d{2}-\d{2}-/.test(post.id || "")) {
    contentType = matchesAny(normalizeText(post.title), classificationRules.contentType.seriesKeywords)
      ? "theory"
      : "daily";
  }

  const market = Object.entries(classificationRules.normalizedTags.market)
    .filter(([, needles]) => matchesAny(combined, needles))
    .map(([tag]) => tag);

  const setup = Object.entries(classificationRules.normalizedTags.setup)
    .filter(([, config]) => {
      const candidateMatch = (config.candidates || []).some((candidate) => candidates.includes(candidate));
      return candidateMatch || matchesAny(combined, config.needles || []);
    })
    .map(([tag]) => tag);

  const topic = Object.entries(classificationRules.normalizedTags.topic)
    .filter(([, needles]) => matchesAny(combined, needles))
    .map(([tag]) => tag);

  let level = classificationRules.defaults.level;
  if (matchesAny(combined, classificationRules.normalizedTags.level.beginner)) {
    level = "beginner";
  } else if (matchesAny(combined, classificationRules.normalizedTags.level.advanced)) {
    level = "advanced";
  }

  return {
    content_type: contentType,
    market: market.length > 0 ? market : [classificationRules.defaults.market],
    setup: setup.length > 0 ? setup : [classificationRules.defaults.setup],
    topic: topic.length > 0 ? topic : [classificationRules.defaults.topic],
    level,
  };
}
